### ISSUE TITLE
What and where in the code the filtering is happening for variants?

### ISSUE THREAD

**Issue by NLSVTN** [broadinstitute/hail-elasticsearch-pipelines#222]
Hi,

We noticed that the number of variants in our VCF is a bit more than the number of rows in the generated MatrixTable. So, does it mean that the pipeline omits some variants? For instance, in one of our VCF we have **382425** but the number of rows in the generated MatrixTable is **356136**. We are interested to know why that happens and where in the code this filtering is happening. I am talking about luigi pipeline:

https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/master/luigi_pipeline/seqr_loading.py#L87

**Comment by hanars** [comment]
Can you proved the command you are executing? 

@mike-w-wilson any idea what might be causing that discrepancy?

**Comment by NLSVTN** [comment]
Just running the luigi pipeline, thats it. It runs fine, generating .mt file, then loading the .mt to Elasticsearch. I don't understand why command should matter at all. We are launching SeqrVCFToMTTask luigi task with such parameters:

SeqrVCFToMTTask(source_paths=["s3://.../CVD.vcf.gz"], dest_path=s3://.../CVD.mt, genome_version=38, array_elements_required=False, vep_runner=VEP, reference_ht_path=s3://.../reference_data_grch38.ht, clinvar_ht_path=s3://.../clinvar.GRCh38.ht, hgmd_ht_path=s3://.../hgmd_hg38.ht, sample_type=WGS, validate=False, dataset_type=VARIANTS, remap_path=, subset_path=)

**Comment by NLSVTN** [comment]
Not fully sure but could we lose variants on importing **vcf** due to some **array_elements_required** parameter?..

**Comment by NLSVTN** [comment]
Or invalid loci: https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/e2a7e08b356ce70af314507908c7df4a9f128d7b/luigi_pipeline/lib/hail_tasks.py#L87

It would be nice if the pipeline would issue some warning message in that case..

**Comment by NLSVTN** [comment]
I tested the parameter skip_invalid_loci setting it to True/False and it does not affect the number of MatrixTable rows, so its not this issue.

**Comment by NLSVTN** [comment]
Same for **array_elements_required** one

**Comment by NLSVTN** [comment]
We did a test run and determined that the line that reduces the number of variants is the following one (.e.g. 382 425 -> 376 133 for one of our dataset):

https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/master/luigi_pipeline/seqr_loading.py#L62

Why does it happen?

For us though the number is also reduced at the line that we currently reasoned to add (but thinking of removing now) that is run also in the script that I added:

https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/5f85a9618ea2ebb8163466db204a413e787d0b3b/hail_scripts/v02/update_reference_data_in_existing_index.py#L65

Basically when there are duplicated loci the pipeline fails if deduplication is not added before line **62** in **seqr_loading.py**. Before I fixed it just by removing duplicate loci but that may not be right. What would you propose to do, how are you handling such a case?

**Comment by NLSVTN** [comment]
The error that happens when we don't remove duplicated loci:

```
Error summary: HailException: RVD error! Keys found out of order:
  Current key:  { locus: { contig: chr1, position: 38046636 }, alleles: [2; GGCC, G] }
  Previous key: { locus: { contig: chr1, position: 38046636 }, alleles: [2; GGCCGCC, G] }
This error can occur after a split_multi if the dataset
contains both multiallelic variants and duplicated loci.
```

**Comment by mike-w-wilson** [comment]
HI @NLSVTN , how are you generating your VCF? The above error happens when the VCF has two distinct multiallelic variants with the same locus. If you do not want to deduplicate, adding `permit_shuffle=True` to `hl.split_multi_hts` here: https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/e2a7e08b356ce70af314507908c7df4a9f128d7b/luigi_pipeline/seqr_loading.py#L96 will shuffle the MT so that the keys are ordered.

**Comment by NLSVTN** [comment]
I am not sure since I am not responsible for this part, but we just talked with the guy who works more closely and he says that its a common thing that happens in highly variable regions and that Hail 0.1 pipeline was somehow fine with that but Hail 0.2 version started giving the error. Thank you for specifying the solution! We are going to try it.

We are still wondering why the function here:

https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/master/luigi_pipeline/seqr_loading.py#L62

Removes some variants?

**Comment by mike-w-wilson** [comment]
Is it possible there are * alleles within the VCF? The `hl.split_multi_hts` will remove star alleles. If so, could you try adding `keep_star=True` in the same line I mentioned above to see if the variant count increases in the MT? 
```
return hl.split_multi_hts(mt.annotate_rows(locus_old=mt.locus, alleles_old=mt.alleles), keep_star=True, permit_shuffle=True)
 ```

**Comment by SimonLi5601** [comment]
> HI @NLSVTN , how are you generating your VCF? The above error happens when the VCF has two distinct multiallelic variants with the same locus. If you do not want to deduplicate, adding `permit_shuffle=True` to `hl.split_multi_hts` here:
> 
> https://github.com/broadinstitute/hail-elasticsearch-pipelines/blob/e2a7e08b356ce70af314507908c7df4a9f128d7b/luigi_pipeline/seqr_loading.py#L96
> 
> will shuffle the MT so that the keys are ordered.

Hi Wilson, thanks for your reply. I was wondering how do you deal with multiallelic variants in your case if it seems you are not using permit_shuffle by default? 
We just tried permit_shuffle=True, however, the job was terminated without giving any specific information. 
In the stderr, it shows "returned non-zero exit status 137. Command exiting with ret '1' "
In the stdout, the last few lines are 
[Stage 14:>                                                         (0 + 0) / 1]
[Stage 14:>                                                         (0 + 1) / 1]2021-03-15 21:03:02 Hail: INFO: Ordering unsorted dataset with network shuffle

**Comment by SimonLi5601** [comment]
Another thing confused me that our vcf is already normalized, we don't have multiallelic with ; in a variant, the entries looks like below,   
chr1	38046636	.	GGCCGCC	G
chr1	38046636	.	GGCCGCC	GGCC
chr1	38046636	.	GGCCGCC	GGCCGCCGCC
chr1	38046640	.	G	A
chr1	38046640	.	G	*

**Comment by mike-w-wilson** [comment]
Hi Simon, 

Our VCFs only have a single locus for multiallelics that are then split in the pipeline. In our experience, the hail error you are seeing is a result of having a multiallelic and a duplicate locus. I'm curious if hail is interpreting the alleles differently than we expect since you said the VCF does not have any multiallelics. Could you try reading in the VCF and then filtering to alleles greater than 2? 
```
mt = hl.import_vcf(path to your vcf)
mt1 = mt.filter_rows(hl.len(mt.alleles) >2)
mt1.count()
```
If the count is greater than zero, hail thinks there is a multiallelic. Could you also send along the version of hail your pipeline is on?

When you add `permit_shuffle=True`, it is triggering a shuffle based on the this message `[Stage 14:> (0 + 1) / 1]2021-03-15 21:03:02 Hail: INFO: Ordering unsorted dataset with network shuffle`. This is a compute intensive step and I'm guessing it is failing because of machine requirements. What type of machine are you running the pipeline on? Also, the last line in the last comment `chr1 38046640 . G *` is a star allele and it will be filtered by our pipeline which I believe will account for some of the variant differences. Could you add `keep_star=True` and see if the MT counts match the VCF?

**Comment by NLSVTN** [comment]
> You were right, it was AWS compute resources issue, we increased them, and it worked fine this time.

> We verified that with `keep_star=True ` number of variants stays the same, so it must be it.

> We also ran the code you proposed regarding multiallelic variants and got the result - (0, 804), so there are none.

> The version of `Hail` is `0.2.57`
